### PR TITLE
Make all network services use network configuration

### PR DESCRIPTION
**Pull request by valib**
Formatted PR #1095 and this PR replace it.

### CODE REVIEW

**Comment by Nadahar** [comment]
@valib I've squashed your commit c9bea11e47bdf3d8abd76e5eaa34ed25cf5ad6c3 into @zero0cool0's ff3421d7c0c26fa997e9094da0374ed9d9f4449e. The new commit is 0922b12e7543869a0f9a739127a3f9d2d727ba96,

I have started to try to address the problems with this implementation, but I don't have time to finish it now.

**Comment by Nadahar** [comment]
I've committed some changes for this. I'm still short on time and have only very briefly tested it. The commit message explains the changes.

Many of the bugs/issues fixed here is inherited from ```HTTPServer.start()``` so corresponding fixes should be implemented there as well. It's a bit more complicated there however.

**Comment by Nadahar** [comment]
@UniversalMediaServer/developers This is a bit of a nightmare - I guess there's a reason why nobody has wanted to touch the network configuration for so long. The web interface itself is easy enough, but for it to have any point I figured that the UPnP configuration should adhere to the configuration as well, and this is where it all starts to get complicated.

Cling is a chapter in itself, I THINK I have that covered now but I'm not quite sure if the changes I made apply both to Cling's UPnP ports and Cling's HTTP server or not. Actually I'm not sure if Cling spawns a HTTP server or not. Any help with this is welcome.

The real nightmare is the "internal" UPnP classes, they are a mess of another dimension. Not only do I have to control what socket they bind to (that in itself is achievable), but I have to make them restartable. This doesn't seem to ever have been a thought, and the shutdown mechanisms are appalling not to say non-existing. The general disregard for interrupts thoughout the code is depressing, and I can't really see any clear strategy for closing down threads. It doesn't help to call ```thread.interrupt()``` when the thread code itself ignores interrupts... and we have blocking sockets that doesn't even respond to interrupts. In all, it's a mess.

This isn't completed by any means, I pushed it simply to let you see a bit of what I'm trying to do. I don't know if this will be successful, and we'll need to test it well. To having any hope of ever completing this, I have to try to look another way regarding a lot of issues found along the way like thread safety, generics etc. Imo the internal UPnP stack should be removed and Cling be used to handle all of this. The current "hybrid solution" doesn't deserve to be called a solution...

**Comment by Nadahar** [comment]
@UniversalMediaServer/developers This turned out to be a big one. This was far from easy, and I can't be sure that I haven't created some new bugs. There are a couple of areas I'm particularly "worried" about:
* I have removed the exclusion of virtual network adapters both in our "internal" UPnP and Cling. From what I could find the reason for this exclusion was simply that some OS X users running some VM's on their desktop OS (VMWare Workstation and similar) had some issues if these virtual adapters were used for multicasting. As I see it, that's not a good enough reason to exclude them. Excluding them means that UPnP won't work properly from any VM either running on a virtualisation server or alongside a desktop OS. It also means that even if the issue with the virtual adapters on OS X is fixed (which it might be already) UPnP still won't work for the VM adapters, neither will it on other platforms without the bug. I think the only correct thing to do is to remove the exclusion and maybe make a note in the documentation that if OS X users have issues with this, they should configure their network to use a specified network interface.
* I have implemented a slightly different multicast socket binding logic, also for Cling. Special arrangements have been made for Windows since the "proper" binding is reported to not work correctly for Windows. One multicast listening socket (and thread) is created PER active interface. This could solve current issues with hosts having multiple network interfaces, but it could also introduce unknown side effects.

There has been a lot of changes here and I expect this to be a bit hard to review. I've tried to split the commits into "logical units of change", even though in reality most of this was done without commiting as it is all very entangled. I've also tried to make descriptive commit messages, but I'm sure there are more that can be said about them. Please don't hesitate to ask if anything is unclear. I might add further details later, but I know from experience that it won't be read anyway so it might be better to just answer any questions you might have.

This will need a lot of testing with different network configurations, it might simply be that this should be released as a beta to gather feedback.

**Comment by atamariya** [comment]
@Nadahar Quick question - did you account for accessing the web socket via multiple interfaces? The problem I'd noticed was that the multicast discovery works. But since web socket used to run on a particular socket, future communication for media used to fail.

**Comment by Nadahar** [comment]
@atamariya Yes, both the "main" HTTP interface and the web interface should reply or multiple interfaces now.

**Comment by atamariya** [comment]
@Nadahar The UPnP discovery returns a device XML (PMS.xml). This has a web URL which is used by the client for future communication. So if the client is in 192.168.0.x, the URL should be in the same subnet. Else the communication fails as no route is found.

**Comment by Nadahar** [comment]
@atamariya I'm aware of that aspect, and I haven't addressed that. This doesn't only apply to the "device" but to every other link/address sent. All the URLs for content would need the same "address aware" logic.

My goal has been to limit this PR to the network configuration itself. I've bumped into several more or less related issues that I've left alone, simply because it would otherwise become huge and untestable/unmergable.

If the client has a route to the other network it should work as it is. If not, if will fail. I'm not sure what address UMS "picks" as it's reference address. From what I've seen in the Cling code, Cling handles this. It figures out which of its own IP addresses is on the same subnet as the client, and then substitutes it's outgoing "reference" address accordingly. From what I can see, that can also fail if there is routing involved, but I haven't studied that in detail. In any case, routing multicast is too complicated for all but the very few, so I'm guessing that is a very rare scenario.

As I see it, UMS' internal UPnP handling should be removed and device and services configured via Cling instead. That would take care of this (maybe except for routed networks) "for free".

**Comment by Nadahar** [comment]
I just tested this on a Ubuntu 12.04 LTS VM and it seems to work fine except that if detects an "unknown renderer" on 127.0.0.1. I'm not sure why, as loopback interfaces are supposed to be excluded.

**Comment by Nadahar** [comment]
I just tested this on OS X 10.8.5, all seems to work fine - no loopback device discovered here. Both of these tests are with a single network card and IP address, but that's my primary concern since that's all the current implementation can handle anyway.

**Comment by Nadahar** [comment]
Same test on Windows 10, no apparant issues.

**Comment by Nadahar** [comment]
Tested on Windows XP 32 bit, no issues.

File: src/main/java/net/pms/network/HTTPServer.java [hunk 1]
```diff
@@ -40,35 +43,97 @@
...
+				socketAddress = new InetSocketAddress(port);
+			}
+		} else {
+			socketAddress = tmpSocketAddress;
+		}
+
+		if (StringUtils.isBlank(tmpHostname) && socketAddress != null) {
+			if (socketAddress.getAddress().isAnyLocalAddress()) {
+				try {
+					tmpHostname = InetAddress.getLocalHost().getHostName();
```
Comment (valib) [inline src/main/java/net/pms/network/HTTPServer.java]: @Nadahar host name doesn't work for my Panasonic TV. This TV understand only the IP address.

File: src/main/java/net/pms/network/HTTPServer.java [hunk 2]
```diff
@@ -40,35 +43,97 @@
...
+					tmpHostname = InetAddress.getLoopbackAddress().getHostName();
+				}
+			} else {
+				tmpHostname = socketAddress.getHostString();
+			}
+		}
+		hostname = tmpHostname;
 	}
 
 	public String getURL() {
```
Comment (valib) [inline src/main/java/net/pms/network/HTTPServer.java]: @Nadahar maybe this should be changed to 

    public String getURL() {
		return "http://" + socketAddress.getAddress().toString() + ":" + port;
	}

Comment (Nadahar) [inline src/main/java/net/pms/network/HTTPServer.java]: @valib I just pushed a commit that takes care of the hostname/IP address issue. 

Are you describing another problem in the comment above a different problem?

**Comment by valib** [comment]
@Nadahar great. Now it works 👍

File: src/main/java/net/pms/network/HTTPServer.java (hunk 2 above)
Comment (valib) [inline src/main/java/net/pms/network/HTTPServer.java]: @Nadahar that problem was related to the hostname using and it was only the description how it  has an influence to the UMS renderers recognition. Now you fixed it and the problem is gone.

**Comment by SubJunk** [comment]
Wow lots of work went into this. Thanks :) Will test on Windows 7 when I get a chance

**Comment by SubJunk** [comment]
But if you're confident that the testing has already been comprehensive enough, and if a code review is done, don't wait for me to test :)

**Comment by Nadahar** [comment]
@SubJunk I'm developing on Windows 7, so it fundamentally works on Windows 7. However, what mostly needs to be tested here is different network configurations, so anyones test can reveal something (like @valib without local name resolution).

I've found the reason for the "localhost" unknown renderer on Ubuntu. It was Cling that was "discovered", the ```isSelf``` definitions in the request handlers just got a bit more complicated now that we listen to multiple interfaces with the HTTP server. The reason it only happened on Ubuntu is that Debian (Ubuntu upstream) has decided to have a "default IP" of ```127.0.1.1```, and the existing ```isSelf``` definitions didn't see that as a local address. By definition anything ```127.*``` is local.

It's such "side effects", things that aren't really related to the changed code but behave unexpected under new circumstances, we need to find.

**Comment by zero0cool0** [comment]
For what it's worth, I've tested this on a dual-interfaced Ubuntu 16.04 LTS using `network_interface=<internal interface>` and it behaves as expected. As far as I can tell, interface bindings are IP address specific for ports 5001, 9001 and 33439 (which I suspect is somehow linked UPNP?). More importantly, none of the external interfaces is bound (also not the loopback interface 127.0.0.1 if that is of interest). This is in-line with what I originally wanted to achieve with this PR.

**Comment by Nadahar** [comment]
@zero0cool0 Thank you for testing 👍 As long as you bind to an interface or IP I think this works well. My testing has shown some issues when not binding (that is, trying to serve all interfaces) on Linux, I'm trying to figure it out.

It might be that I have to bite the bullet and implement dynamic "reference" address as @atamariya described, even though I wasn't ready for that extra work in this PR.

**Comment by Nadahar** [comment]
Just tested on Arch, same results as Ubuntu (detecting itself as a renderer on loopback before last commit).

**Comment by Nadahar** [comment]
@UniversalMediaServer/developers I just pushed my current "work in progress" on this. To fix the problems detected, I see no other solution than to make a "proper" network configuration where individual network interfaces and/or IP addresses are selectable.

This means that I have to do GUI redesign, and as much as I hate that I've started to do it. It took me a long time to get my custom ```JTree``` up and working properly. 

This is in no way complete, in fact the network configuration itself is quite bugged in this state, I'm only focusing on the GUI for now. I'd like to get feedback on what I'm doing on the "General" tab, as GUI work is so time consuming that I don't want to continue if I have to redo it later.

Currently I have some test code in there that generates some random "sub ip addresses" to show in the tree. This is for testing the tree only, and won't be there in the final code. Changes aren't saved to the configuration either at this stage.

What is new/I want feedback on are:
* Grouping the network configuration the way I've done it. 
* The interface/address selection tree
* The new ```DefaultTextField``` I've created and is currently used for hostname and port. The way it works is that it's like a normal text field except that when it's blank, the default is shown in a "muted color". 

Also, I'm not very happy with the icons used in the interface/address selection tree, but I decided to use them for now and move on, because I have a tendency of being stuck with such things for a loong time. Any suggestions there is very welcome.

**Comment by Nadahar** [comment]
What an overwhelming response - doesn't anybody have ANY views on the current GUI "solution"?

**Comment by valib** [comment]
@Nadahar I am not looking for the changes all over the time so sometimes the respond would take a time. This looks good but I am not sure if there will be any other IP addresses for the particular interface.

**Comment by Nadahar** [comment]
@valib Normally there won't be multiple IP addresses per interface, but an interface can have multiple IP addresses (and sub interfaces) so I've tried to model the configuration to cater for all possibilities.

**Comment by SubJunk** [comment]
Testing this is on my todo list :)

**Comment by Nadahar** [comment]
@SubJunk There's not much to test at this stage, a quick look at the GUI changes is all that's needed atm. I need to know it the direction is endorsed or not before I put more work into it, GUI simply takes soo much time to get right.

**Comment by SubJunk** [comment]
@Nadahar that looks awesome, the GUI seems like a big improvement to me

**Comment by Nadahar** [comment]
I've reset the branch to it's original state, and this work will continue in #1195 instead. I vote for not merging this as it is, as it won't work if network configuration is changed. #1195 will solve what this PR intended.

**Comment by SubJunk** [comment]
Closing this as it was superseded by #1195

**Pull request closed without merging**